markets/perps-market/contracts/storage/InterestRate.sol

```diff
@@ -0,0 +1,104 @@
+//SPDX-License-Identifier: MIT
+pragma solidity >=0.8.11 <0.9.0;
+
+import {SafeCastU256, SafeCastU128, SafeCastI256} from "@synthetixio/core-contracts/contracts/utils/SafeCast.sol";
+import {DecimalMath} from "@synthetixio/core-contracts/contracts/utils/DecimalMath.sol";
+import {GlobalPerpsMarket} from "./GlobalPerpsMarket.sol";
+import {Position} from "./Position.sol";
+import {GlobalPerpsMarketConfiguration} from "../storage/GlobalPerpsMarketConfiguration.sol";
+
+library InterestRate {
+    using DecimalMath for uint256;
+    using DecimalMath for uint128;
+    using SafeCastU128 for uint128;
+    using SafeCastU256 for uint256;
+    using SafeCastI256 for int256;
+    using GlobalPerpsMarket for GlobalPerpsMarket.Data;
+    using Position for Position.Data;
+    // 4 year average which includes leap
+    uint256 private constant AVERAGE_SECONDS_PER_YEAR = 31557600;
+
+    bytes32 private constant _SLOT_INTEREST_RATE =
+        keccak256(abi.encode("io.synthetix.perps-market.InterestRate"));
+
+    struct Data {
+        uint256 interestAccrued; // per $1 of OI
+        uint128 interestRate;
+        uint256 lastTimestamp;
+    }
+
+    function load() internal pure returns (Data storage interestRate) {
+        bytes32 s = _SLOT_INTEREST_RATE;
+        assembly {
+            interestRate.slot := s
+        }
+    }
+
+    function update() internal returns (uint128 newInterestRate, uint256 currentInterestAccrued) {
+        Data storage self = load();
+
+        (
+            uint128 lowUtilizationInterestRateGradient,
+            uint128 interestRateGradientBreakpoint,
+            uint128 highUtilizationInterestRateGradient
+        ) = GlobalPerpsMarketConfiguration.loadInterestRateParameters();
+
+        // if no interest parameters are set, interest rate is 0 and the interest accrued stays the same
+        if (
+            lowUtilizationInterestRateGradient == 0 ||
```

Comment:
should be `AND` and not `OR`

Reply:
hmmm, initially i did think if any of these are 0, it should ignore it, but you're right, maybe low could be 0 and the other two could be set which might be pretty common.  changing to AND, thx